### DISCUSSION TITLE
How do i allocate array on free store using new?

### DISCUSSION THREAD

**Discussion by jituderaps** [hsutter/cppfront#740] [Q&A] [answered]
[https://cpp2.godbolt.org/z/s3vqrW6v1](url)
```
main: () = {
    pi: std::unique_ptr<int>;

    pi = new<std::array<int, 5>>(35);

    std::cout << pi* << "\n";
}
```

**Comment by hsutter** [comment, marked as answer]
Thanks, good question!

You could always do this:

```
//  Example 1 - works, but does extra init+copy work
main: () = {
    pi: std::unique_ptr<std::array<int,5>>;
    pi = new<std::array<int, 5>>();
    pi* = (1, 2, 3, 4, 5);
    for pi* do (i) std::cout << i;
}
```

Or this:

```
//  Example 2 - also works, but does extra init+copy work
main: () = {
    pi: std::unique_ptr<std::array<int,5>>;
    pi = new<std::array<int, 5>>( :std::array = (1, 2, 3, 4, 5) );
    for pi* do (i) std::cout << i;
}
```

But both of those involve default-initializing all the array elements, and then deep-copying them all.

I've now enabled this to work as expected (commit e6a1cc363c54e5bcbf0381cd3fade6ef86c33f9d)-- I thought this would already have worked because of P0960, but apparently not so I implemented the `if constexpr( requires` check myself to use list initialization if available:

```
//  Example 3 - now works, without init+copy overhead
main: () = {
    pi := new<std::array<int, 5>>(1, 2, 3, 4, 5);  // direct list init
    for pi* do (i) std::cout << "(i)$\n";
}
```

Thanks!

**Reply by JohelEGP** [reply]
The problem is that parentheses don't perform brace elision:
- <https://quuxplusone.github.io/blog/2022/06/03/aggregate-parens-init-considered-kinda-bad/#curly-braced-initialization-list>.
- Paragraph specifically about `std::array`: <https://quuxplusone.github.io/blog/2022/06/03/aggregate-parens-init-considered-kinda-bad/#the-reason-is-that-b-1-2-3-remai>.

Clang and GCC's error messages suggest that: <https://compiler-explorer.com/z/z3fchxzhY>.
